### ISSUE TITLE
after some time viewobj stops working

### ISSUE THREAD

**Issue by idavydov** [jalvesaq/Nvim-R#761]
Hi,

I typically load a large Rmd report and work on it for a while. I noticed that after some time I am not able to view objects using `<localleader>rv` (`RAction("viewobj")` or via `:call RAction("viewobj")`); nothing happens after RAction is called. 

Print functionality still works (either via `<localleader>rp` or via `:call`).

Since I do not see anything in the `:messages`, I am not sure how to debug this problem; `:RDebugInfo` also doesn't have anything suspicious.

**Comment by jalvesaq** [comment]
You can put in your `~/.Rprofile`:

```r
options(nvimcom.verbose = 4)
```

Then, you will see messages sent to and coming from `nvimcom`.

**Comment by idavydov** [comment]
thanks, @jalvesaq .

I confirm that `options(nvimcom.verbose = 4)` generates some extra output in the R terminal.

E.g., when I press `<localleader>rp` I see:
```
> print(xxx)
# A tibble: 169 × 4
...
# ℹ 159 more rows
# ℹ Use `print(n = ...)` to see more rows
nvimcom_nvimclient(RTaskCompleted()): '10101' (10101)
> 
```

However when I press `<localleader>rv` (or run `:call RAction("viewobj")`) absolutely nothing happens.

If I run this in the terminal, viewobj works (a new tab is created):
```
> nvimcom:::nvim_viewobj(xxx)
nvimcom_nvimclient(RViewDF('pa_corresp', 'tabnew')): '10101' (10101)
nvimcom_nvimclient(RTaskCompleted()): '10101' (10101)
```

**Comment by idavydov** [comment]
I also noticed that in this state "<localleader>rh" doesn't open a help tab (also nothing happens).

**Comment by jalvesaq** [comment]
Maybe another plugin is mapping `<LocalLeader>rv`... You can check this with the command:

```vim
:nmap
```

**Comment by idavydov** [comment]
I do not think it's binding, since the `:call RAction("viewobj")` also does not do anything (`RAction("print")` however works fine).

I checked `:nmap`, `<LocalLeader>rv` is assigned to `:call RAction("viewobj")`.

**Comment by jalvesaq** [comment]
`RAction("print")` works because the command is sent to R Console, not to nvimcom, but it's not a problem of communication from Nvim-R to nvimcom because you can see the messages when `nvimcom.verbose=4`. What `RAction("viewobj")` and `RAction("help")` have in common is that both trigger the same procedure:

1. Neovim sends a message to nvimcom.
2. Nvimcom creates a file in Nvim-R's temporary directory.
3. Nvimcom sends a message to Neovim.
4. Neovim reads the file created by nvimcom: `Rinsert` for "viewobj" and `Rdoc` for "help".

The files `Rinsert` and `Rdoc` are replaced each time you run the commands that create them, but they should be deleted only on the `VimLeave` event.

**Comment by jalvesaq** [comment]
To check if Neovim is really receiving messages from nvimcom, it is necessary to uncomment the first line of the function `ROnJobStdout()` at `R/nvimrcom.vim` and delete the string "DEBUG:". It will become:

```vim
call writefile(a:data, "/dev/shm/nclientserver_stdout", "a")
```
Then, in another terminal or in another Tmux pane, you would have to use `tail` to follow the output:

```sh
tail -f /dev/shm/nclientserver_stdout
```

**Comment by idavydov** [comment]
thanks, @jalvesaq 

Do I understand correctly that since I don't see any output in R console (with `nvimcom.verbose = 4`) this means that either neovim doesn't send a message or the communication is somehow broken?

And I checked, when I reach this state when object view doesn't work, there's no output in nclientserver_stdout; but my understanding is that this is expected since even the first step doesn't happen.

**Comment by jalvesaq** [comment]
Yes, you are correct. I can only imagine that `nvimcom` was unloaded, the `nclientserver` application was killed, or the communication was blocked by a firewall. This command will show whether `nclientserver` is running or not:

```sh
ps -Af | grep nclientserver | grep -v grep
```

**Comment by idavydov** [comment]
hm, it seems that `nclientserver` is still running `username 1563498 1563091  0 11:13 ?        00:00:00 /xxx/R/x86_64-pc-linux-gnu-library/4.3.0-foss/nvimcom/bin/nclientserver` :-(

there's no easy way to switch to a unix socket to check if it's firewall, correct?

**Comment by idavydov** [comment]
is there maybe a way to start `nclientserver` and connect via netcat and check if connection breaks at some point?

**Comment by jalvesaq** [comment]
This should work (replace 10101 with the actual `nclientserver` port):

```sh
echo "Hello" | nc -u -q 0 -v 127.0.0.1 10101
```

**Comment by jalvesaq** [comment]
I replaced UDP connection with TCP on the [remote](https://github.com/jalvesaq/Nvim-R/tree/remote) branch to fix another issue. I guess this change will make no difference here but, perhaps, you could try it anyway since I can't replicate this issue.

**Comment by idavydov** [comment]
just an update from my side that for the last couple of days I was unable to reproduce this problem.

as soon as I will, I will test if the connection is alive and report back.

**Comment by jalvesaq** [comment]
Thanks for the feedback! Are you using the remote branch? I will merge the remote branch into the master branch in a few days.

**Comment by idavydov** [comment]
No, I'm still on the master branch waiting for the issue to reproduce, so I can diagnose what's with nclientserver

**Comment by idavydov** [comment]
It has been a couple of weeks when I couldn't reproduce the issue anymore. I think it makes sense to close it now. I'll reopen in case the issue reappears.
